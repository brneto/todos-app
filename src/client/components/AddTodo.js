import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { useAddTodo } from '../hooks/todo';

const Title = styled.h1`
  position: absolute;
  top: -155px;
  width: 100%;
  color: rgba(175, 47, 47, .15);
  font-size: 100px;
  font-weight: 100;
  text-align: center;
  text-rendering: optimizeLegibility;
`;

const Input = styled.input`
  position: relative;
  box-sizing: border-box;
  width: 100%;
  margin: 0;
  padding: 16px;
  border: none;
  background: rgba(0, 0, 0, .003);
  color: inherit;
  font-family: inherit;
  font-size: 24px;
  font-weight: inherit;
  line-height: 1.4em;
  box-shadow: inset 0 -2px 1px rgba(0, 0, 0, .03);

  ::placeholder {
    color: #e6e6e6;
    font-style: italic;
    font-weight: 300;
  }
`;

export function AddTodoPresentation({ dispatch }) {
  const input = useAddTodo(dispatch);
  return (
    <header>
      <Title>todos</Title>
      <Input name="todo" placeholder="What needs to be done?" {...input} />
    </header>
  );
}

AddTodoPresentation.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

export default connect()(AddTodoPresentation);
