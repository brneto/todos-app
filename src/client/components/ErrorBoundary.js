import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { getFilterPath } from '../api';

const
  OnError = styled.label`
    margin-left: 1em;

    & > button {
      margin: 1em .5em;
    }
  `;

class ErrorBoundary extends React.Component {
  static propTypes = {
    children: PropTypes.element.isRequired,
    onRetry: PropTypes.func.isRequired,
  };

  state = { error: null, prevFilter: getFilterPath() };

  // Update state so the next render will show the fallback UI.
  static getDerivedStateFromError = error => ({ error });

  componentDidUpdate() {
    const
      { prevFilter } = this.state,
      filter = getFilterPath();

    if (filter !== prevFilter)
      this.setState({ error: null, prevFilter: filter });
  }

  retryHandler = () => this.setState({ error: null }) || void this.props.onRetry();

  render() {
    const {
      props: { children },
      state: { error },
      retryHandler,
    } = this;

    return error ? (
      <OnError>Could not fetch todos. {error.message}
        <button onClick={retryHandler}>Retry</button>
      </OnError>
    ) : children;
  }
}

export default ErrorBoundary;
