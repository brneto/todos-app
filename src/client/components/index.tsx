export { default as LoadingDots } from './LoadingDots';
export { default as ErrorBoundary } from './ErrorBoundary';
