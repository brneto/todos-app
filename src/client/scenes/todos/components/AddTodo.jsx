import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { connect } from 'react-redux';
import { todos } from './hooks';

const
  Header = styled.header`
    & > h1 {
      position: absolute;
      top: -155px;
      width: 100%;
      color: rgba(175, 47, 47, .15);
      font-size: 100px;
      font-weight: 100;
      text-align: center;
      text-rendering: optimizeLegibility;
    }

    & > input {
      position: relative;
      box-sizing: border-box;
      width: 100%;
      margin: 0;
      padding: 16px;
      border: none;
      background: rgba(0, 0, 0, .003);
      color: inherit;
      font-family: inherit;
      font-size: 24px;
      font-weight: inherit;
      line-height: 1.4em;
      box-shadow: inset 0 -2px 1px rgba(0, 0, 0, .03);

      ::placeholder {
        color: #e6e6e6;
        font-style: italic;
        font-weight: 300;
      }
    }
  `;

  const
    subscribe = connect(),
    propTypes = { dispatch: PropTypes.func.isRequired };

export function AddTodoPresentation({ dispatch }) {
  // Custom hook:
  const input = todos.useAddTodo(dispatch);

  return (
    <Header>
      <h1>todos</h1>
      <input name="todo" placeholder="What needs to be done?" {...input} />
    </Header>
  );
}
AddTodoPresentation.propTypes = propTypes;

export default subscribe(AddTodoPresentation);
