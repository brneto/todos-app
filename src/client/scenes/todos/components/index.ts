export { default as AddTodo, AddTodoPresentation } from './AddTodo';
export { default as VisibleTodoList } from './VisibleTodoList';
export { default as FilterList } from './FilterList';
export { default as TickerTape } from './TickerTape';
