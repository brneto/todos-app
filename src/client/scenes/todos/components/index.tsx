export { default as AddTodo, AddTodoPresentation } from './AddTodo';
export { default as VisibleTodoList } from './VisibleTodoList';
export { default as Footer } from './Footer';
export { default as TickerTape } from './TickerTape';
