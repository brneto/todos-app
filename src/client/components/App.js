import React from 'react';
import styled from 'styled-components';
import AddTodo from './AddTodo';
import VisibleTodoList from './VisibleTodoList';
import Footer from './Footer';
import TickerTape from './TickerTape';

// TODO: Replace styled-components by Emotion
// TODO: https://emotion.sh/docs/introduction
// TODO: styled-components vs emotion
// TODO: https://github.com/jsjoeio/styled-components-vs-emotion/blob/master/README.md
const
  TodoApp = styled.div({
    position: 'relative',
    margin: '130px 0 40px 0',
    background: '#fff',
    boxShadow: [
      '0 2px 4px 0 rgba(0, 0, 0, .2)',
      '0 25px 50px 0 rgba(0, 0, 0, .1)',
    ]
  }),
  App = () => (
    <TodoApp>
      <AddTodo />
      <VisibleTodoList />
      <Footer />
      <TickerTape url="api/sse" />
    </TodoApp>
  );

export default App;
