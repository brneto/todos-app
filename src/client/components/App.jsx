import React from 'react';
import styled from '@emotion/styled';
import AddTodo from './AddTodo';
import VisibleTodoList from './VisibleTodoList';
import Footer from './Footer';
import TickerTape from './TickerTape';

const
  TodoApp = styled.div`
    position: relative;
    margin: 130px 0 40px 0;
    background: #fff;
    box-shadow:
      0 2px 4px 0 rgba(0, 0, 0, .2),
      0 25px 50px 0 rgba(0, 0, 0, .1);
  `;

const
  App = () => (
    <TodoApp>
      <AddTodo />
      <VisibleTodoList />
      <Footer />
      <TickerTape url="api/sse" />
    </TodoApp>
  );

export default App;
