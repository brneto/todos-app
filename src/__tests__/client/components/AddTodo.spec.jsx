import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { AddTodoPresentation } from '../../../client/components/AddTodo';

describe('components/AddTodo', () => {
  it('should call handle methods', () => {
    const dispatch = jest.fn();

    // render the todo form with a fake dispatch
    const props = { dispatch };
    const { getByPlaceholderText, container } = render(
      <AddTodoPresentation {...props} />
    );

    // get the todo input
    const todoInput = getByPlaceholderText('What needs to be done?');
    // Fires onChange event
    fireEvent.change(todoInput, { target: { value: 'change' } });
    // Fires enter onKeyDown event
    fireEvent.keyDown(todoInput, { keyCode: 13 });

    // assert that our fake dispatch was called with the right stuff
    expect(dispatch).toHaveBeenCalledTimes(1);
    // assert if equal to snapshot
    expect(container.firstChild).toMatchSnapshot();
  });
});
