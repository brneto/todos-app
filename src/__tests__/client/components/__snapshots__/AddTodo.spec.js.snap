// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AddTodo should call handle methods 1`] = `
.c0 {
  position: absolute;
  top: -155px;
  width: 100%;
  color: rgba(175,47,47,.15);
  font-size: 100px;
  font-weight: 100;
  text-align: center;
  text-rendering: optimizeLegibility;
}

.c1 {
  position: relative;
  box-sizing: border-box;
  width: 100%;
  margin: 0;
  padding: 16px;
  border: none;
  background: rgba(0,0,0,.003);
  color: inherit;
  font-family: inherit;
  font-size: 24px;
  font-weight: inherit;
  line-height: 1.4em;
  box-shadow: inset 0 -2px 1px rgba(0,0,0,.03);
}

.c1::-webkit-input-placeholder {
  color: #e6e6e6;
  font-style: italic;
  font-weight: 300;
}

.c1::-moz-placeholder {
  color: #e6e6e6;
  font-style: italic;
  font-weight: 300;
}

.c1:-ms-input-placeholder {
  color: #e6e6e6;
  font-style: italic;
  font-weight: 300;
}

.c1::placeholder {
  color: #e6e6e6;
  font-style: italic;
  font-weight: 300;
}

<header>
  <h1
    class="c0"
  >
    todos
  </h1>
  <input
    class="c1"
    name="todo"
    placeholder="What needs to be done?"
    value=""
  />
</header>
`;
